### PR TITLE
[GLUTEN-3511][CORE] Support to add custom aggregate functions for extension

### LINKED ISSUE #3511: [CORE]Support to add custom aggregate functions for extension
### Description

Support to add custom aggregate functions:

When transforming `HashAggregateExec`, it needs to get attributes from the aggregate function, but if there are some custom aggregate functions, it can not handle these custom aggregate functions when transforming.

Reuse `ExpressionExtensionTrait` to support this feature, add a new function called `getAttrsForExtensionAggregateExpr` to handle this.

### PR DESCRIPTION
**Pull request by zzcclp**
## What changes were proposed in this pull request?
When transforming HashAggregateExec, it needs to get attributes from the aggregate function, but if there are some custom aggregate functions, it can not handle these custom aggregate functions when transforming.

Reuse ExpressionExtensionTrait to support this feature, add a new function called getAttrsForExtensionAggregateExpr to handle this.

Close #3511.

(Fixes: #3511)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala [hunk 1]
```diff
@@ -447,7 +447,19 @@ abstract class HashAggregateExecBaseTransformer(
     var resIndex = index
     val mode = exp.mode
     val aggregateFunc = exp.aggregateFunction
-    aggregateFunc match {
+    val reIndex = aggregateFunc match {
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala:450]: What is the difference of `reIndex` and `resIndex`? Seems to be easily mixed up.

File: cpp-ch/local-engine/Parser/aggregate_function_parser/CommonAggregateFunctionParser.cpp [hunk 2]
```diff
@@ -62,4 +62,6 @@ REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER(Ntile, ntile, ntile)
 REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER(PercentRank, percent_rank, percent_rank)
 REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER(CumeDist, cume_dist, cume_dist)
 
+// Only for ut
+REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER(CustomSum, custom_sum, sum)
```
Comment (baibaichen) [inline cpp-ch/local-engine/Parser/aggregate_function_parser/CommonAggregateFunctionParser.cpp:66]: Moving `REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER` into header  and then moving `REGISTER_COMMON_AGGREGATE_FUNCTION_PARSER(CustomSum, custom_sum, sum)` into seperated cpp.

Refer to [cpp-ch/local-engine/Parser/example_udf/myAdd.cpp](https://github.com/oap-project/gluten/blob/main/cpp-ch/local-engine/Parser/example_udf/myAdd.cpp) and #1589

File: backends-clickhouse/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CustomSum.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.trees.TreePattern.{SUM, TreePattern}
+import org.apache.spark.sql.catalyst.trees.UnaryLike
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+/** Copy from Vanilla Spark 3.2 */
```
Comment (PHILO-HE) [inline backends-clickhouse/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CustomSum.scala:28]: Nit:
could you revise the comment to let others known it is ported from Spark 3.2 `Sum`? If there is any code change, please also clarify. Thanks!

**Review by PHILO-HE — COMMENTED** [review]
Looks good! Just a trivial comment.

File: gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala (hunk 1 above)
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala:450]: Done

File: cpp-ch/local-engine/Parser/aggregate_function_parser/CommonAggregateFunctionParser.cpp (hunk 2 above)
Comment (zzcclp) [reply cpp-ch/local-engine/Parser/aggregate_function_parser/CommonAggregateFunctionParser.cpp:66]: Done

File: backends-clickhouse/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CustomSum.scala (hunk 3 above)
Comment (zzcclp) [reply backends-clickhouse/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CustomSum.scala:28]: Done

**Comment by zzcclp** [comment]
@PHILO-HE @rui-mo @baibaichen comments fix, please help to review again, thanks.

**Review by PHILO-HE — APPROVED** [review]
LGTM!

**Review by rui-mo — APPROVED** [review]

**Pull request merged**